### ISSUE TITLE
Slow queries on larger result sets

### ISSUE THREAD

**Issue by ceit-each** [mikkopaderes/ember-cloud-firestore-adapter#229]
Hi there,

I am new to this addon after working with emberfire for a good while (looking to move on due to lack of maintenance). Appreciate the work and the addon looks great!
However, one thing I've noticed is that when using `query` with a result set in the 1000s there is a fairly significant difference in the amount of time it takes for both addons to return the documents.

In my simple test, to retrieve 2000 documents using `this.store.query`, both addons took roughly the following amount of time to return:
ember-cloud-firestore-adapter: ~ 18 seconds
emberfire: ~ 4-5 seconds

I also would periodically come across the following error:
```
[2022-01-11T14:25:46.640Z]  @firebase/firestore: Firestore (9.6.2): FirebaseError: [code=resource-exhausted]: Too many outstanding requests.
[2022-01-11T14:25:46.641Z]  @firebase/firestore: Firestore (9.6.2): Using maximum backoff delay to prevent overloading the backend.
```

I have only briefly looked at the code of the addon, but is the individual fetching of docs in `query() -> findQueryRecords()` in [adapters/cloud-firestore-modular.ts](https://github.com/mikkopaderes/ember-cloud-firestore-adapter/blob/e4431a32812fe0cf99c8076952abfaac2ebbbcfa/addon/adapters/cloud-firestore-modular.ts#L246) the cause of these issues?

**Comment by mikkopaderes** [comment]
Yes, there are individual fetches after a query but if you notice, the query uses `onSnapshot` and does the individual fetching inside of it. Then at the very end, it does a unsubscribe to it. 

Because we kept a listener running, the individual fetches will attempt to fetch from the cache making it resolve the request significantly faster. 

To be honest, it's been a while since I did this simple benchmarking and I'm not sure if it's still the case or if Firestore has changed that behavior already. Maybe it's time for me to check that out soon.

**Comment by mikkopaderes** [comment]
I've checked this again, and it does seem to not act as I would've remembered then. I'm gonna try to revisit the implementation to improve ng querying.

**Comment by ceit-each** [comment]
Great, thanks for looking into it!

**Comment by mikkopaderes** [comment]
#237 introduced the refactor
